In [1]:
import pandas as pd
import numpy as np

data = pd.read_csv(r"C:\Users\rohit\OneDrive\Desktop\Modules\MLPractice/diabetes.csv")
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [2]:
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [4]:
from sklearn.model_selection import train_test_split

X = data.drop("Outcome", axis = "columns")
y = data["Outcome"]

print(X)
print(y)

from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X=scaler.fit_transform(X)

X_train_full, X_test, y_train_full, y_test = train_test_split(X, y, train_size = 0.5, random_state = 0)


     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [5]:
data.shape

(768, 9)

In [6]:
X_train_full.shape

(384, 8)

In [7]:
X_train, X_val, y_train, y_val = train_test_split(X_train_full, y_train_full, test_size=0.3, random_state=42)


In [8]:
X_train.shape

(268, 8)

In [9]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
log_reg.fit(X_train,y_train)


LogisticRegression()

In [10]:
from sklearn.pipeline import Pipeline
from sklearn import svm
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

svm_model = svm.SVC(kernel='rbf')
svm_model.fit(X_train, y_train)


SVC()

In [11]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)

RandomForestClassifier()

In [12]:
from sklearn.model_selection import cross_val_score

cross_val_score(log_reg, X_train, y_train, cv=3, scoring="accuracy")


array([0.75555556, 0.71910112, 0.79775281])

In [13]:
cross_val_score(svm_model, X_train, y_train, cv=3, scoring="accuracy")


array([0.72222222, 0.71910112, 0.75280899])

In [14]:
cross_val_score(rfc, X_train, y_train, cv=3, scoring="accuracy")


array([0.72222222, 0.69662921, 0.75280899])

In [15]:
from sklearn.metrics import accuracy_score

y_pred = log_reg.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.7916666666666666


In [16]:
y_pred = svm_model.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.7421875


In [17]:
y_pred = rfc.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.7447916666666666


In [18]:
import tensorflow as tf
from tensorflow import keras

model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[8, ]),
    keras.layers.Dense(10, activation="relu"),
    keras.layers.Dense(5, activation="relu"),
    keras.layers.Dense(1, activation="softmax")
])


In [19]:
model.layers

In [20]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 8)                 0         
                                                                 
 dense (Dense)               (None, 10)                90        
                                                                 
 dense_1 (Dense)             (None, 5)                 55        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 151
Trainable params: 151
Non-trainable params: 0
_________________________________________________________________


In [21]:
hidden1 = model.layers[1]
#hidden2 = model.layers[2]
weights, biases = hidden1.get_weights()
weights

array([[-0.2691693 , -0.14000472, -0.04063636,  0.56182635, -0.37622738,
         0.47982097, -0.1359129 , -0.11598805,  0.15433133, -0.01918304],
       [ 0.02284652, -0.02404761, -0.25724143, -0.2880568 ,  0.36500758,
        -0.10655883,  0.12225008, -0.12616599,  0.4147873 ,  0.008129  ],
       [-0.19309026,  0.2662251 ,  0.45296288, -0.27457526,  0.05892497,
        -0.35323155,  0.48692882,  0.04670447,  0.2606824 , -0.09722814],
       [-0.17472982,  0.54640067,  0.0786041 , -0.34025297, -0.09637457,
        -0.5600826 ,  0.34859365, -0.17387569,  0.04530245, -0.427407  ],
       [ 0.1457153 ,  0.1586991 ,  0.44109023,  0.05602509,  0.30317503,
         0.3905161 , -0.06022161, -0.37674493,  0.5168892 ,  0.19891852],
       [-0.19894826,  0.55030715,  0.49260128,  0.45882607,  0.37024146,
         0.4804157 , -0.32041994,  0.52741337,  0.02195233, -0.41010493],
       [-0.4831101 , -0.18816057, -0.4258451 , -0.37132382, -0.33986437,
        -0.1160664 ,  0.2383765 ,  0.25440168

In [22]:
biases

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)

In [23]:
weights.shape

(8, 10)

In [24]:
biases.shape

(10,)

In [25]:
model.compile(loss="binary_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [26]:
history = model.fit(X_train, y_train, epochs=30,
                    validation_data=(X_val, y_val))

Epoch 1/30
9/9 [==============================] - 2s 56ms/step - loss: 0.9011 - accuracy: 0.3657 - val_loss: 0.9389 - val_accuracy: 0.3362
Epoch 2/30
9/9 [==============================] - 0s 8ms/step - loss: 0.8543 - accuracy: 0.3657 - val_loss: 0.8942 - val_accuracy: 0.3362
Epoch 3/30
9/9 [==============================] - 0s 8ms/step - loss: 0.8160 - accuracy: 0.3657 - val_loss: 0.8565 - val_accuracy: 0.3362
Epoch 4/30
9/9 [==============================] - 0s 9ms/step - loss: 0.7843 - accuracy: 0.3657 - val_loss: 0.8282 - val_accuracy: 0.3362
Epoch 5/30
9/9 [==============================] - 0s 9ms/step - loss: 0.7601 - accuracy: 0.3657 - val_loss: 0.8032 - val_accuracy: 0.3362
Epoch 6/30
9/9 [==============================] - 0s 8ms/step - loss: 0.7386 - accuracy: 0.3657 - val_loss: 0.7836 - val_accuracy: 0.3362
Epoch 7/30
9/9 [==============================] - 0s 8ms/step - loss: 0.7224 - accuracy: 0.3657 - val_loss: 0.7668 - val_accuracy: 0.3362
Epoch 8/30
9/9 [=================

In [27]:
model.evaluate(X_test, y_test)

12/12 [==============================] - 0s 3ms/step - loss: 0.6154 - accuracy: 0.3411


[0.6153542399406433, 0.3411458432674408]

In [28]:
y_pred = log_reg.predict(X_test)
Logistic_regression_classifier = print("Accuracy of Logistic regression= " ,accuracy_score(y_test, y_pred))

y_pred = svm_model.predict(X_test)
SVM_classifier = print("Accuracy of SVM = " ,accuracy_score(y_test, y_pred))

y_pred = rfc.predict(X_test)
RandomForest_classifier = print("Accuracy of RandomForest Classifier = " ,accuracy_score(y_test, y_pred))

Accuracy of Logistic regression=  0.7916666666666666
Accuracy of SVM =  0.7421875
Accuracy of RandomForest Classifier =  0.7447916666666666


Here I noticed that Logistic Regression has the highest accuracy with a 79% accuracy
With DNN, we can actually see the change in the accuracy values with regards to the validation loss with each epoch. 